redun/executors/k8s.py

```diff
@@ -0,0 +1,1174 @@
+    """
+    Parse task error from s3 scratch path.
+    """
+    assert job.task
+
+    error_path = aws_utils.get_job_scratch_file(s3_scratch_prefix, job, s3_utils.S3_SCRATCH_ERROR)
+    error_file = File(error_path)
+
+    if not job.task.script:
+        # Normal Tasks (non-script) store errors as Pickled exception, traceback
+        # tuples.
+        if error_file.exists():
+            error, error_traceback = pickle.loads(cast(bytes, error_file.read("rb")))
+        else:
+            error = K8SError("Exception and traceback could not be found for K8S Job.")
+            error_traceback = Traceback.from_error(error)
+    else:
+        # Script task.
+        if error_file.exists():
+            error = ScriptError(cast(bytes, error_file.read("rb")))
+        else:
+            error = K8SError("stderr could not be found for K8S Job.")
+        error_traceback = Traceback.from_error(error)
+
+    return error, error_traceback
+
+
+def parse_task_logs(k8s_job_id: str, namespace, max_lines: int = 1000) -> Iterator[str]:
+    """
+    Iterates through most recent logs of an K8S Job.
+    """
+
+    lines_iter = iter_k8s_job_log_lines(
+        k8s_job_id,
+        namespace,
+        reverse=True,
+    )
+
+    i = islice(lines_iter, 0, max_lines)
+    lines = reversed(list(i))
+
+    if next(lines_iter, None) is not None:
+        yield "\n*** Earlier logs are truncated ***\n"
+    yield from lines
+
+
+def k8s_describe_jobs(
+    job_names: List[str], namespace: str
+) -> typing.List[kubernetes.client.V1Job]:
+    """
+    Returns K8S Job descriptions from the AWS API.
+    """
+    api_instance = k8s_utils.get_k8s_batch_client()
+    responses = []
+    for job_name in job_names:
+        api_response = api_instance.read_namespaced_job(job_name, namespace=namespace)
+        responses.append(api_response)
+    return responses
+
+
+def iter_k8s_job_status(job_ids, namespace):
+    """Wraps k8s_describe_jobs in a generator"""
+    for job in k8s_describe_jobs(job_ids, namespace):
+        yield job
+
+
+def iter_log_stream(
+    job_id: str,
+    namespace: str,
+    reverse: bool = False,
+) -> Iterator[str]:
+    """
+    Iterate through the events of a K8S log.
+    """
+
+    job = k8s_describe_jobs([job_id], namespace)[0]
+    api_instance = k8s_utils.get_k8s_core_client()
+    label_selector = f"job-name={job.metadata.name}"
+    api_response = api_instance.list_pod_for_all_namespaces(label_selector=label_selector)
+
+    if not api_response.items:
+        lines = [f"Cannot find pod for job {job.metadata.name}"]
+    else:
+        # Use most recent pod for job.
+        pod = api_response.items[-1]
+        name = pod.metadata.name
+        namespace = pod.metadata.namespace
+        log_response = api_instance.read_namespaced_pod_log(name, namespace=namespace)
+        lines = log_response.split("\n")
+
+        # Use latest container status.
```

Comment:
I added pod-level logs here to help debug cases where a pod didn't even fully start due to OOM. `state.message` contained very help info to include.